### PR TITLE
SONARJAVA-1155 - S2325 FP when creating non static anonymous inner cl…

### PR DESCRIPTION
**Pull request by DidierBesset**
…ass accessing members and false negative with method belonging to same hierarchy

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 1]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+    void setNonStaticReference() {
+      nonStaticReference = true;
+    }
+
+    boolean hasNonStaticReference() {
+      return nonStaticReference;
+    }
+
+    void checkSymbol(Symbol symbol) {
+      if (nonStaticReference || symbol == methodSymbol || symbol.isStatic()) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 2]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+        setNonStaticReference();
+      }
+    }
+
+    private boolean isConstructor(Symbol symbol) {
+      return "<init>".equals(symbol.name());
+    }
+
+    private static boolean hasLocalAccess(Symbol scope, Symbol symbol) {
+      if (scope == symbol) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 3]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+    public void checkNonConstructor(final Symbol symbol, final Symbol scopeOwner) {
+      if (scopeOwner.isMethodSymbol()) {
+        return;
+      }
+      if (hasLocalAccess(methodScopeOwner, scopeOwner)) {
+        setNonStaticReference();
+      }
+    }
+
+    private boolean isConstructor(Symbol symbol) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: MAJOR Make "isConstructor" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 4]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+      return "<init>".equals(symbol.name());
+    }
+
+    private static boolean hasLocalAccess(Symbol scope, Symbol symbol) {
+      if (scope == symbol) {
+        return true;
+      }
+      if (scope.isTypeSymbol() && symbol.isTypeSymbol()) {
+        // FIXME use erasure until Parameterized types are correctly supported!
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 5]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+    private static boolean isConstructor(Symbol symbol) {
+      return "<init>".equals(symbol.name());
+    }
+
+    private static boolean hasLocalAccess(Symbol scope, Symbol symbol) {
+      if (scope.equals(symbol)) {
+        return true;
+      }
+      if (scope.isTypeSymbol() && symbol.isTypeSymbol()) {
+        // FIXME use erasure until Parameterized types are correctly supported!
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/test/files/checks/StaticMethodCheck.java [hunk 6]
```diff
@@ -1,8 +1,13 @@
 package foo;
 
+import trials.Converter;
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/StaticMethodCheck.java]: These 2 imports are useless.

File: java-checks/src/test/files/checks/StaticMethodCheck.java [hunk 7]
```diff
@@ -11,6 +16,18 @@
 
   public Utilities() {
   }
+  
+  public void registerPrimitives(final boolean type) {
+    register(Boolean.TYPE, new Toto());
+  }
+  
+  public void register(final Class<?> clazz, final Object converter) {
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/StaticMethodCheck.java]: these 3 methods are ``public`` do you really need them?

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 8]
```diff
@@ -62,8 +66,7 @@
 
   private JavaFileScannerContext context;
   private Deque<Symbol> outerClasses = new LinkedList<>();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: ``outerClasses`` should be removed, as you don't use it anymore.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 9]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+    }
+
+    @CheckForNull
+    private static Symbol scopeOwner(Symbol symbol) {
+      Symbol owner = symbol.owner();
+      while (owner != null) {
+        if (owner.isPackageSymbol() || owner.isTypeSymbol() || owner.isMethodSymbol()) {
+          break;
+        }
+        owner = owner.owner();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: Not covered and probably impossible to cover, as the owner of a symbol can hardly be something else than a package, a type or a method. I feel that the whole method is useless. ``methodScopeOwner`` should simply be ``methodOwner``.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 10]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
   public void visitIdentifier(IdentifierTree tree) {
     super.visitIdentifier(tree);
+    if ("class".equals(tree.name()) || methodReferences.isEmpty()) {
+      return;
+    }
     Symbol symbol = tree.symbol();
-    if (!atLeastOneReference.isEmpty() && !atLeastOneReference.peek() && !currentMethod.peek().equals(symbol)  && referenceInstance(symbol)) {
-      atLeastOneReference.pop();
-      atLeastOneReference.push(Boolean.TRUE);
+    final MethodReference currentMethod = methodReferences.peek();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: remove ``final`` modifier.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 11]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+
+    boolean hasNonStaticReference() {
+      return nonStaticReference;
+    }
+
+    void checkSymbol(Symbol symbol) {
+      if (nonStaticReference || methodSymbol.equals(symbol) || symbol.isStatic()) {
+        return;
+      }
+      final Symbol scopeOwner = scopeOwner(symbol);
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: remove ``final`` modifier.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 12]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+      if (isConstructor(symbol)) {
+        checkConstructor(scopeOwner);
+      } else if (scopeOwner != null) {
+        checkNonConstructor(symbol, scopeOwner);
+      }
+    }
+
+    public void checkConstructor(final Symbol constructorClass) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: should be ``private``.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 13]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
+      if (!constructorClass.isStatic()) {
+        Symbol methodPackage = getPackage(methodScopeOwner);
+        Symbol constructorPackage = getPackage(constructorClass);
+        if (Objects.equals(methodPackage, constructorPackage) && !constructorClass.owner().isPackageSymbol()) {
+          setNonStaticReference();
+        }
+      }
+    }
+
+    public void checkNonConstructor(final Symbol symbol, final Symbol scopeOwner) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: should be ``private``.

File: java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java [hunk 14]
```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
...
     }
   }
 
-  private boolean referenceInstance(Symbol symbol) {
-    return symbol.isUnknown() || (!symbol.isStatic() && fromInstance(symbol.owner()));
+  @Override
+  public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+    if (tree.expression().is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) tree.expression();
+      final Symbol owner = identifier.symbol().owner();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java]: remove ``final`` modifier.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**